Null AutogradMeta optimization

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28620 Merge Tensor and Variable.
* **#28610 Null AutogradMeta optimization**
* #28609 Delete dead TensorImpl::detach_autograd_meta
* #28602 Remove unused gradient_edge argument from make_variable_view
* #28593 Add an AutogradMeta factory.
* #28601 Unfriend Variable factory functions.
* #28592 Move all autograd_meta_ manipulating operations out-of-line.
* #28543 Test TensorTypeSet instead of autograd_meta_ for variable-ness.

The basic idea is, in some cases where we stored a pointer to a full AutogradMeta object, instead store a nullptr. We let a nullptr represent a default-constructed AutogradMeta object, and simply populate it with a real AutogradMeta if there is ever a situation where we need to modify it.

The primary technical contrivance in this diff is I have to use AutogradMetaFactory to lazily initialize the AutogradMeta, as it is not available in the dynamic library that TensorImpl is in. (I spent a while trying to put them in the same compilation unit, but gave up in the end as it pushed us over the Windows linking binary size limit. Eep.)

Some other notes:
- `set_autograd_meta` now unconditionally turns a tensor into a variable. I audited all call sites and observed there are no occurrences where nullptr is passed (after this patch, there are now!)
- `copy_tensor_metadata` is updated to unconditionally preserve the VariableTensorId-ness of the destination tensor. I think this is the more correct semantics; we can't do the old semantics anymore.
- There's a bunch of places in the API where we return const references to objects. This is pretty weird to me, but I didn't feel like cleaning it up. But sometimes I don't conveniently have something that's the right lifetime, so I introduced a number of singletons to handle this correctly.

You might wonder why I'm doing the optimization before the variable-tensor dynamic merge. The reason is simple: this change is semantics preserving, while variable-tensor dynamic merge is not. So it is easier to get right, and prevents us from regressing performance if we do it the other way.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171162](https://our.internmc.facebook.com/intern/diff/D18171162)

ezyang committed:
Null AutogradMeta optimization

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Null AutogradMeta optimization"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Null AutogradMeta optimization"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

yf225 commented on c10/core/TensorImpl.cpp:
```diff
@@ -225,25 +227,27 @@ at::DataPtr PlacementDeleteContext::makeDataPtr(
+  TORCH_INTERNAL_ASSERT(type_set_.has(TensorTypeId::VariableTensorId), "set_requires_grad is not implemented for Tensor");
+  if (!requires_grad && !autograd_meta_) return;
+  if (!autograd_meta_) autograd_meta_ = impl::GetAutogradMetaFactory()->make();
+  autograd_meta_->set_requires_grad(requires_grad, this);
```
Discussed with @ezyang offline, and it seems that we have three possible states of a Variable after this PR:
1) Variable with `autograd_meta_ == nullptr`
2) Variable with `autograd_meta_` default constructed
3) Variable with `autograd_meta_` fully populated

I think it would be all good if at the end of the merge we only have 1) and 3), which corresponds to `requires_grad` being false / true, respectively. It would also be awesome to have a comment about the intermediate state :D

albanD commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -125,8 +133,13 @@ DifferentiableViewMeta::~DifferentiableViewMeta() {
+
 const std::shared_ptr<Node>& Variable::grad_fn() const {
   if (is_view()) {
+    // NB: is_view() ==> get_autograd_meta()
```
Does this means that we are sure that if `is_view() == true` then `get_autograd_meta != nullptr`?

ezyang replied:
Yep that's right

albanD commented on torch/csrc/autograd/variable.h:
```diff
@@ -623,16 +640,25 @@ inline Variable Variable::detach() const {
+  if (get_autograd_meta()) {
+    return get_autograd_meta()->output_nr_;
+  } else {
+    return 0;
```
Have you considered this to be an error? This should never be called on a Tensor or default constructed autograd meta no?

ezyang replied:
An error might be appropriate. But for this diff, I opted for the semantics preserving option, which is zero (the default value).

albanD commented on c10/core/TensorImpl.cpp:
```diff
@@ -29,16 +29,18 @@ const char * const TensorImpl::err_msg_tensor_metadata_change_not_allowed =
 at::Tensor& TensorImpl::grad() {
-  if (autograd_meta()) {
-    return autograd_meta()->grad();
+  if (type_set_.has(TensorTypeId::VariableTensorId)) {
```
This could be wrapped in a nice function to make code more readable?
I'm fine if you don't do it because it will go away with the next commit.

ezyang replied:
Yes, it goes away next commit :>

albanD commented on c10/core/TensorImpl.cpp:
```diff
@@ -225,25 +227,27 @@ at::DataPtr PlacementDeleteContext::makeDataPtr(
 }
 
 void TensorImpl::set_autograd_meta(std::unique_ptr<c10::AutogradMetaInterface> autograd_meta) {
+  // NB: autograd_meta may be null!  That just means it's the default
```
Maybe add a comment here that reflects the discussion above about the autograd meta that can be in 3 possible states?

albanD reviewed (approved):
lgtm, comments about doc mostly.

ezyang committed:
Update on "Null AutogradMeta optimization"


The basic idea is, in some cases where we stored a pointer to a full AutogradMeta object, instead store a nullptr. We let a nullptr represent a default-constructed AutogradMeta object, and simply populate it with a real AutogradMeta if there is ever a situation where we need to modify it.

The primary technical contrivance in this diff is I have to use AutogradMetaFactory to lazily initialize the AutogradMeta, as it is not available in the dynamic library that TensorImpl is in. (I spent a while trying to put them in the same compilation unit, but gave up in the end as it pushed us over the Windows linking binary size limit. Eep.)

Some other notes:
- `set_autograd_meta` now unconditionally turns a tensor into a variable. I audited all call sites and observed there are no occurrences where nullptr is passed (after this patch, there are now!)
- `copy_tensor_metadata` is updated to unconditionally preserve the VariableTensorId-ness of the destination tensor. I think this is the more correct semantics; we can't do the old semantics anymore.
- There's a bunch of places in the API where we return const references to objects. This is pretty weird to me, but I didn't feel like cleaning it up. But sometimes I don't conveniently have something that's the right lifetime, so I introduced a number of singletons to handle this correctly.

You might wonder why I'm doing the optimization before the variable-tensor dynamic merge. The reason is simple: this change is semantics preserving, while variable-tensor dynamic merge is not. So it is easier to get right, and prevents us from regressing performance if we do it the other way.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

gchanan reviewed (approved):
I have some questions and small suggestions.

Most of those are about pre-existing code though, so I don't think this needs to be held up.

ezyang committed:
Update on "Null AutogradMeta optimization"


The basic idea is, in some cases where we stored a pointer to a full AutogradMeta object, instead store a nullptr. We let a nullptr represent a default-constructed AutogradMeta object, and simply populate it with a real AutogradMeta if there is ever a situation where we need to modify it.

The primary technical contrivance in this diff is I have to use AutogradMetaFactory to lazily initialize the AutogradMeta, as it is not available in the dynamic library that TensorImpl is in. (I spent a while trying to put them in the same compilation unit, but gave up in the end as it pushed us over the Windows linking binary size limit. Eep.)

Some other notes:
- `set_autograd_meta` now unconditionally turns a tensor into a variable. I audited all call sites and observed there are no occurrences where nullptr is passed (after this patch, there are now!)
- `copy_tensor_metadata` is updated to unconditionally preserve the VariableTensorId-ness of the destination tensor. I think this is the more correct semantics; we can't do the old semantics anymore.
- There's a bunch of places in the API where we return const references to objects. This is pretty weird to me, but I didn't feel like cleaning it up. But sometimes I don't conveniently have something that's the right lifetime, so I introduced a number of singletons to handle this correctly.

You might wonder why I'm doing the optimization before the variable-tensor dynamic merge. The reason is simple: this change is semantics preserving, while variable-tensor dynamic merge is not. So it is easier to get right, and prevents us from regressing performance if we do it the other way.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171162](https://our.internmc.facebook.com/intern/diff/D18171162)

[ghstack-poisoned]

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -227,25 +229,34 @@ at::DataPtr PlacementDeleteContext::makeDataPtr(
 void TensorImpl::set_requires_grad(bool requires_grad) {
-  TORCH_INTERNAL_ASSERT(autograd_meta(), "set_requires_grad is not implemented for Tensor");
-  autograd_meta()->set_requires_grad(requires_grad, this);
+  TORCH_INTERNAL_ASSERT(type_set_.has(TensorTypeId::VariableTensorId), "set_requires_grad is not implemented for Tensor");
```
Ya, I agree it's user triggerable, although the assert is being removed next commit so it's all moot XD

ezyang commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -64,6 +67,9 @@ void Variable::detach_() {
   if (is_view()) {
     AT_ERROR("Can't detach views in-place. Use detach() instead");
   }
+  // I think the choice here is conservative.  But there is some weirdness going
```
Didn't notice anything. I was auditing Variable methods to see where I need to place materialize, and when I got to `detach_` I thought to myself, "well, morally, we should be able to just set `autograd_meta_ = false` here and we're done." But then I noticed we're not clearing out things like the debugName and hooks... were they supposed to be cleared in `detach_`? I didn't want to work out the answer to this question in this PR.

ezyang replied:
(I'll improve the comment.)

ezyang commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -64,6 +67,9 @@ void Variable::detach_() {
   }
+  // I think the choice here is conservative.  But there is some weirdness going
+  // on: why aren't, e.g., hooks cleared in this case?
+  materialize_autograd_meta();
```
Sure

ezyang replied:
My new idea is to just make `materialize_autograd_meta` return an `AutogradMeta*`

gchanan replied:
I think that's slightly worse, because making it part of the get API explicitly forces you to think about it (I don't even need to read the docs really, it's obvious from the parameter name what's going on).

But returning is an improvement.

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -1553,8 +1561,28 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
+  //
+  // Note that we don't enforce the invariant that if the AutogradMeta is
+  // default constructed, it is nullptr (to do this, we'd have to continuously
+  // check if an AutogradMeta became, by mutation, it's equal to the default
```
or just equal XD

ezyang committed:
Update on "Null AutogradMeta optimization"


The basic idea is, in some cases where we stored a pointer to a full AutogradMeta object, instead store a nullptr. We let a nullptr represent a default-constructed AutogradMeta object, and simply populate it with a real AutogradMeta if there is ever a situation where we need to modify it.

The primary technical contrivance in this diff is I have to use AutogradMetaFactory to lazily initialize the AutogradMeta, as it is not available in the dynamic library that TensorImpl is in. (I spent a while trying to put them in the same compilation unit, but gave up in the end as it pushed us over the Windows linking binary size limit. Eep.)

Some other notes:
- `set_autograd_meta` now unconditionally turns a tensor into a variable. I audited all call sites and observed there are no occurrences where nullptr is passed (after this patch, there are now!)
- `copy_tensor_metadata` is updated to unconditionally preserve the VariableTensorId-ness of the destination tensor. I think this is the more correct semantics; we can't do the old semantics anymore.
- There's a bunch of places in the API where we return const references to objects. This is pretty weird to me, but I didn't feel like cleaning it up. But sometimes I don't conveniently have something that's the right lifetime, so I introduced a number of singletons to handle this correctly.

You might wonder why I'm doing the optimization before the variable-tensor dynamic merge. The reason is simple: this change is semantics preserving, while variable-tensor dynamic merge is not. So it is easier to get right, and prevents us from regressing performance if we do it the other way.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171162](https://our.internmc.facebook.com/intern/diff/D18171162)

[ghstack-poisoned]

ezyang committed:
Update on "Null AutogradMeta optimization"


The basic idea is, in some cases where we stored a pointer to a full AutogradMeta object, instead store a nullptr. We let a nullptr represent a default-constructed AutogradMeta object, and simply populate it with a real AutogradMeta if there is ever a situation where we need to modify it.

The primary technical contrivance in this diff is I have to use AutogradMetaFactory to lazily initialize the AutogradMeta, as it is not available in the dynamic library that TensorImpl is in. (I spent a while trying to put them in the same compilation unit, but gave up in the end as it pushed us over the Windows linking binary size limit. Eep.)

Some other notes:
- `set_autograd_meta` now unconditionally turns a tensor into a variable. I audited all call sites and observed there are no occurrences where nullptr is passed (after this patch, there are now!)
- `copy_tensor_metadata` is updated to unconditionally preserve the VariableTensorId-ness of the destination tensor. I think this is the more correct semantics; we can't do the old semantics anymore.
- There's a bunch of places in the API where we return const references to objects. This is pretty weird to me, but I didn't feel like cleaning it up. But sometimes I don't conveniently have something that's the right lifetime, so I introduced a number of singletons to handle this correctly.

You might wonder why I'm doing the optimization before the variable-tensor dynamic merge. The reason is simple: this change is semantics preserving, while variable-tensor dynamic merge is not. So it is easier to get right, and prevents us from regressing performance if we do it the other way.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171162](https://our.internmc.facebook.com/intern/diff/D18171162)

[ghstack-poisoned]

ezyang committed:
Update on "Null AutogradMeta optimization"


The basic idea is, in some cases where we stored a pointer to a full AutogradMeta object, instead store a nullptr. We let a nullptr represent a default-constructed AutogradMeta object, and simply populate it with a real AutogradMeta if there is ever a situation where we need to modify it.

The primary technical contrivance in this diff is I have to use AutogradMetaFactory to lazily initialize the AutogradMeta, as it is not available in the dynamic library that TensorImpl is in. (I spent a while trying to put them in the same compilation unit, but gave up in the end as it pushed us over the Windows linking binary size limit. Eep.)

Some other notes:
- `set_autograd_meta` now unconditionally turns a tensor into a variable. I audited all call sites and observed there are no occurrences where nullptr is passed (after this patch, there are now!)
- `copy_tensor_metadata` is updated to unconditionally preserve the VariableTensorId-ness of the destination tensor. I think this is the more correct semantics; we can't do the old semantics anymore.
- There's a bunch of places in the API where we return const references to objects. This is pretty weird to me, but I didn't feel like cleaning it up. But sometimes I don't conveniently have something that's the right lifetime, so I introduced a number of singletons to handle this correctly.

You might wonder why I'm doing the optimization before the variable-tensor dynamic merge. The reason is simple: this change is semantics preserving, while variable-tensor dynamic merge is not. So it is easier to get right, and prevents us from regressing performance if we do it the other way.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18171162](https://our.internmc.facebook.com/intern/diff/D18171162)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 9b4b618:
* 0/1 flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=9b4b618256383c770fc8ab04efb817b7d4fa15f9).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).
